module/component/component.go

```diff
@@ -110,43 +109,23 @@ func RunComponent(ctx context.Context, componentFactory ComponentFactory, handle
 			default:
 				panic(fmt.Sprintf("invalid error handling result: %v", result))
 			}
-		case <-done:
-			// Without this additional select, there is a race condition here where the done channel
-			// could have been closed as a result of an irrecoverable error being thrown, so that when
-			// the scheduler yields control back to this goroutine, both channels are available to read
-			// from. If this last case happens to be chosen at random to proceed instead of the one
-			// above, then we would return as if the component shutdown gracefully, when in fact it
-			// encountered an irrecoverable error.
-			select {
-			case err := <-irrecoverableErr:
-				switch result := handler(err); result {
-				case ErrorHandlingRestart:
-					continue
-				case ErrorHandlingStop:
-					return err
-				default:
-					panic(fmt.Sprintf("invalid error handling result: %v", result))
-				}
-			default:
-			}
-
-			// Similarly, the done channel could have closed as a result of the context being canceled.
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			default:
-			}
-
-			// clean completion
-			return nil
+		} else if util.CheckClosed(ctx.Done()) {
```

Comment:
nit:
```suggestion
		} else if ctx.Err() != nil {
```

Reply:
`util.CheckClosed` is still useful for situations where we only have a done channel, but when we actually have a context, checking `ctx.Err()` should be the preferred way of doing it because it avoids having to do a select.

Reply:
updated in 8e4ac7e